core/org.wso2.carbon.user.core/src/main/java/org/wso2/carbon/user/core/config/UserStoreConfigXMLProcessor.java

```diff
@@ -430,143 +427,142 @@ private static PrivateKey getPrivateKey() {
      * @throws BadPaddingException
      * @throws IllegalBlockSizeException
      */
-    private String decryptProperty(String propValue)
-            throws NoSuchPaddingException, NoSuchAlgorithmException, NoSuchProviderException,
-            org.wso2.carbon.user.api.UserStoreException, InvalidKeyException, BadPaddingException,
-            IllegalBlockSizeException {
+    private String decryptProperty(String propValue) throws CryptoException {
 
-        Cipher keyStoreCipher;
+        CryptoService cryptoService = UserStoreMgtDataHolder.getInstance().getCryptoService();
         String cipherTransformation = System.getProperty(CIPHER_TRANSFORMATION_SYSTEM_PROPERTY);
         byte[] cipherTextBytes = Base64.decode(propValue.trim());
+        String algorithm = null;
+        byte[] decryptedValue;
+        boolean isInternalKeyStoreEncryptionEnabled = false;
+        boolean isSymmetricKeyEncryptionEnabled = false;
+        ServerConfigurationService config =
+                UserStoreMgtDSComponent.getServerConfigurationService();
+        if (config != null) {
+            String encryptionKeyStore = config.getFirstProperty(ENCRYPTION_KEYSTORE);
 
-        privateKey = (privateKey == null) ? getPrivateKey() : privateKey;
-        if (privateKey == null) {
-            throw new org.wso2.carbon.user.api.UserStoreException(
-                    "Private key initialization failed. Cannot decrypt the userstore password.");
+            if (INTERNAL_KEYSTORE.equalsIgnoreCase(encryptionKeyStore)) {
+                isInternalKeyStoreEncryptionEnabled = true;
+            }
+            String cryptoProvider = config.getFirstProperty(CRYPTO_PROVIDER);
+            if (SYMMETRIC_KEY_CRYPTO_PROVIDER.equalsIgnoreCase(cryptoProvider)) {
+                isSymmetricKeyEncryptionEnabled = true;
+            }
         }
 
-        if(cipherTransformation != null) {
-            // extract the original cipher if custom transformation is used configured in carbon.properties.
-            CipherHolder cipherHolder = cipherTextToCipherHolder(cipherTextBytes);
-            if (cipherHolder != null) {
-                // cipher with meta data.
+        if (isInternalKeyStoreEncryptionEnabled && isSymmetricKeyEncryptionEnabled) {
+
+            throw new CryptoException(String.format("Userstore encryption can not be supported due to " +
+                    "conflicting configurations: '%s' and '%s'. When using internal keystore, assymetric crypto " +
+                    "provider should be used.", INTERNAL_KEYSTORE, SYMMETRIC_KEY_CRYPTO_PROVIDER));
+        } else if (isInternalKeyStoreEncryptionEnabled || isSymmetricKeyEncryptionEnabled) {
```

Comment:
Doesn't this skip the case where internal key store is not configured and symmetric is not enabled, which should switch to primary key store

Reply:
There is another else statement after this else if. It's handled there.
https://github.com/wso2/carbon-kernel/pull/2639/files#diff-889aa9c6d5d53b652ccbcfdb23b953a7R486

Reply:
noted the condition is handled